### ISSUE TITLE
Add side selection to ballot set form

### ISSUE THREAD

**Issue by czlee** [czlee/tabbycat#158]

**Comment by czlee** [comment]
Made a start in 54b5531.

Question: Is it better to swap the order of `<div>`s in the UI, or to swap the speaker name and team name field contents? `SpeakerScore`s and `TeamScore`s are tied to `DebateTeam`s, not positions. That is, you can swap the "positions" in the database without affecting scores. So ideally we would think of fields on the form as being attached to "teams" rather than "sides", as this saves us a lookup on save ("which `DebateTeam` is the aff team?").

This, however, would require a ground-up redesign of the form... this might not actually be that hard, it's mainly just renaming fields, but it would need to be done. Also, I don't know how hard it is to just rearrange HTML elements at will, though I think this is the basis for a lot of interactive stuff?

As an aside, attaching fields to teams rather than sides also makes it easier to generalize the form to BP, if we ever do so :wink:

**Comment by czlee** [comment]
In all fields, `t#` is a team ID (_e.g._ `t24`), `a#` is an adjudicator ID, `s#` is a speaker position (`s1`, `s2`, `s3` for substantives, `s4` for reply).

**Status quo. If field IDs are in terms of aff/neg**, they will look like this:
```
password
choose_sides
motion
aff_motion_veto, neg_motion_veto
aff_speaker_s#, neg_speaker_s#          - returned by position.aff_speaker, position.neg_speaker
aff_score_a#_s#, neg_score_a#_s#        - returned by position.aff_score, position.neg_score
```
`password`, `choose_sides`, and `motion` are team-independent and rendered directly by the template. `aff_motion_veto` and `neg_motion_veto` are rendered directly by the template. The remaining fields are specified by an iterator `form.adj_iter`, that goes like this:
```
    {% for adj in form.adj_iter %}
        {% for position in adj.position_iter %}
            {% if forloop.parentloop.first %}
                {{ position.aff_speaker }}
            {% else %}
                <div type="text" disabled>Copy speaker from first adjudicator<div>        
            {% endif %}
            {{ position.aff_score }}
        {% endfor %}
        {% for position in adj.position_iter %}
            {% if forloop.parentloop.first %}
                {{ position.neg_speaker }}
            {% else %}
                <div type="text" disabled>Copy speaker from first adjudicator<div>        
            {% endif %}
            {{ position.neg_score }}
        {% endfor %}
    {% endfor %}
```


**Proposal. If field IDs are in terms of team ID**, they will look like this:
```
password
choose_sides
motion
motion_veto_t#, motion_veto_t#
speaker_t#_s#, speaker_t#_s#
score_a#_s#, score_a#_s#
```

**Comment by czlee** [comment]
Notes relating to 13a9d48706b73b85c0cd57ce484228c5dbb30922 @philipbelesky:

1. On load, the form checks to see if the sides are already known in the database, and if they are, sets the initial value of the `choose_sides` field to what it finds. I did this so that it will show side information with draft ballots that need to be confirmed (or if for whatever reason a user wants to revisit a ballot). But the JavaScript currently won't check its initial value, and it always hides the score fields, so this needs to be fixed.

    On other note, currently, sides are always known, because we don't actually use `DebateTeam.POSITION_UNALLOCATED`. I think it'd be too much work (for now) to fix this: assumptions that teams always have a side are embedded throughout the code base. So the effect is that the `choose_sides` field will always be initialized.

2. There was an interesting issue where two errors cancelled out. The properties `aff_team`, _etc._ of `Debate` were `cached_property`s, so when you swapped the sides, it would still save scores to the *old* sides, which was "correct" because we were pulling UI elements left and right so that "neg" really meant "aff" and vice versa. I uncached those properties (befa3d8e38faff734d499c80b7c14f212d3c8808) to make them just `property`s, and now it breaks, as expected. It should start working again once speaker dropdown substitution is implemented.

3. Access to the team ID numbers, and the fake speaker selects, was added in 13a9d48706b73b85c0cd57ce484228c5dbb30922. (Nothing else was added in the same commit so it should be easy to spot.)

4. **Known issue.** Whenever you save a new ballot, it *always modifies the sides in the database*. This means that saving a draft ballot modifies sides, but *confirming a ballot does not*, so the sides that will be in the database will be whichever were the last to be entered (not confirmed). This system will break if there are ballots with conflicting information about sides. It's not possible to fix this without moving sides information to `BallotSubmission`, which would be a huge database architectural change.
